### ISSUE TITLE
Encode colors as 32 bit integers by default

### ISSUE THREAD

**Issue by jakirkham** [bokeh/bokeh#5676]
It is not uncommon when working with a lot of data or large images to need many different colors. At present, these colors are packed into a JSON list where each color is a string. The color strings themselves beginning with a `#` and are followed by a 6 digit hexadecimal number. However, this can end up taking up a fair bit of space, which eats into transfer time and/or storage depending on the use case.

Recently PR ( https://github.com/bokeh/bokeh/pull/5544 ) was added to provide support encoding NumPy arrays as `base64` and decoding them as the appropriate typed JavaScript arrays on the other end. Even though there is no additional compression considered beyond this encoding scheme, the result is a drastic reduction in size of data. This makes a big difference in terms of being able to distribute results and, in some cases, loading them. Various array type data are already handled very nicely by this change.

In light of the aforementioned PR, it is worth considering applying a similar strategy to colors. For instance, `Greys256` takes up ~2.5KB in its current form in an HTML document. However, if we use a 256 x 3 `uint8` NumPy array, and encode it using the `base64` scheme from PR ( https://github.com/bokeh/bokeh/pull/5544 ) the same color range takes up ~1KB. In other words, we could see a ~2.5x reduction in size where color palettes are concerned by doing nothing more than allowing NumPy arrays as a color argument and extending the encoding work from PR ( https://github.com/bokeh/bokeh/pull/5544 ) (if needed) to include this case. A rough calculation backs this up as well. Conversion back to an appropriate JavaScript color string is pretty straightforward.

To summarize, it would be really interesting to explore encoding lists of color strings into compact `base64` representations. By more compactly structuring our data, we should be able to more freely represent our data and without worrying about how to share it.

**Comment by bryevdv** [comment]
I do think this is a good idea, but FWIW I think for different reasons. Without more work, we'd have to encode all colors as 32bit integers. That's what BokehJS expects. The palettes themselves are small and may or not even be sent, so the size is not really a consideration to me, at least. But if you are sending CDS columns of colors with thousands of colors, then there could be aa significant *speed* improvement, even if there is not a space improvement.

**Comment by jakirkham** [comment]
> Without more work, we'd have to encode all colors as 32bit integers.

So is this RGBA or ARGB or something like this?

Conversion could be quite fast to go from `uint8` to `uint32` thanks to NumPy's [`view`]( https://docs.scipy.org/doc/numpy/reference/generated/numpy.ndarray.view.html ). Though I'm sure you are already thinking about this. 😄

> But if you are sending CDS columns of colors with thousands of colors, then there could be a significant speed improvement, even if there is not a space improvement.

Depending on the data, it could be getting up there.

In what sense would one see a speed improvement? Load time, transfer time, interactivity, and/or something else?

**Comment by bryevdv** [comment]
> So is this RGBA or ARGB or something like this?

RGBA IIRC

> Conversion could be quite fast to go from uint8 to uint32 thanks to NumPy's view. Though I'm sure you are already thinking about this. 

Right, converting to uint8 on the python side is easy. But BokehJS only knows how to interpret colors as 32bit ints, so if we ship uint8 to BokehJS there would need to be some special case added to BokehJS to handle that

> Depending on the data, it could be getting up there.

In what sense would one see a speed improvement? Load time, transfer time, interactivity, and/or something else?

load time, update time. The base64 encoding isn't particularly smaller (size-wise) than the old JSON encoding (it may even be slightly bigger). But decoding and getting data directly into to typed arrays is *much* faster. As one very extreme example, a 2000x2000 pixel image could take ~50 seconds to render in the notebook. Now it's ~4-5 seconds.

**Comment by jakirkham** [comment]
> Right, converting to uint8 on the python side is easy. But BokehJS only knows how to interpret colors as 32bit ints, so if we ship uint8 to BokehJS there would need to be some special case added to BokehJS to handle that

Ah so I was suggesting that we actually convert `uint8` (and whatever else is supported e.g. hex strings) to `uint32` on the Python side and then pack that in the plot. This was BokehJS has what needs. Plus the array is 1D so there is no need to think about striding, conversion, etc.

**Comment by bryevdv** [comment]
> Ah so I was suggesting that we actually convert uint8 (and whatever else is supported e.g. hex strings) to uint32 on the Python side and then pack that in the plot. This was BokehJS has what needs. Plus the array is 1D so there is no need to think about striding, conversion, etc.

OK, but in that case why not just start with uint32 everywhere? To be clear my idea for this was simply to convert every lists of color strings in `bokeh.palettes` to a corresponding numpy array with `dtype=uint32`.

**Comment by jakirkham** [comment]
From a user standpoint, IMHO it is easier to work with `uint8` arrays when creating a new palette. So I thought it would be nice for the user if we handled RGB or RGBA as `uint8` and did the conversion for them to `uint32`. Does that seem reasonable? Maybe this could be a utility function.

As far as stock palettes, I completely agree that having them be `uint32` out of the box makes sense.

**Comment by jakirkham** [comment]
> > Depending on the data, it could be getting up there.
>>
> > In what sense would one see a speed improvement? Load time, transfer time, interactivity, and/or something else?
>
> load time, update time. The base64 encoding isn't particularly smaller (size-wise) than the old JSON encoding (it may even be slightly bigger). But decoding and getting data directly into to typed arrays is much faster. As one very extreme example, a 2000x2000 pixel image could take ~50 seconds to render in the notebook. Now it's ~4-5 seconds.

No kidding. I have data from before that would sometimes crash my browser and had to be compressed before emailing. Now it loads quickly, is pretty interactive, and I could send 2 or 3 such plots in an email sans compression. Very nice improvements indeed!

Though there is still room for improvement w.r.t load time. So if this change helps with that too, I'm all for it. 👍

**Comment by bryevdv** [comment]
I think we are maybe talking at ends? I am thinking about optimizing the existing palettes, but you are thinking more about how users might create or specify new palettes? If so, sure that could be any number of things.

**Comment by jakirkham** [comment]
I think so. TBH I'd be happy to see improvements on both ends as I use palettes from Bokeh and construct my own.

**Comment by jakirkham** [comment]
Any reason this has been relabeled to a discussion topic? It seemed like feature was an accurate description of this proposed change.

**Comment by bryevdv** [comment]
After some thought I have some doubts about making all the palettes be NumPy arrays on the python. Although NumPy is technically a requirement of Bokeh, it is possible to use Bokeh with plain lists just fine, and some folks who use Bokeh outside of more "science-y" domains probably appreciate this. Making all the palettes be NumPy arrays all the time makes a much harder dependency on NumPy, and I am just not certain of the benefit. Most of the palettes are under 10 values, this is not going to be the case that the binary protocol best serves. 

The original post mentions getting a savings by making a `uint8` encoding in the document itself, but this is not something easily done at all, since everything on the BokehJS side expect 32 bit ints, so there is no way to put 8bit ints into the document without serious changes the BokehJS side. 

There was also mention about making python functions for creating palettes, but not a concrete proposal for what functions and what they do. 

Reading over this I basically I am still unclear on the proposal at this point.

**Comment by jakirkham** [comment]
I don't know that they need to be NumPy arrays. Couldn't they just be lists of integers? How does PR ( https://github.com/bokeh/bokeh/pull/5544 ) handle the conversion? I thought it was already able to handle this list of integers case. Is that not true? 😕

It's true if we have to tack in alpha with every value that this could change things. However, I'm unclear why we need to tack in alpha. It doesn't appear to be included now. Is that an option? 😕

Even if alpha is included, I imagine there is still some savings though probably less. Assuming of course we are comparing not having alpha now (with the hex string) to having alpha in `uint32`. Though that doesn't really feel like an apple-to-apples comparison as we are adding new information.

While the discussion up to this point has focused on the benefits of space storage, we have overlooked a crucial point. Namely it is much easier for users to work with lists of integers when composing a color palette than it is a hex string. For me personally, I had to take on another dependency just to solve this problem, which feels really silly.

In short, this doesn't make NumPy a hard requirement, can work just as well with `uint32`, and will improve usability.

**Comment by bryevdv** [comment]
ping @mattpap this is an old discussion, but to revive it, I'd like to consider/suggest the scoped concrete proposal of:

* allow sending colors as integers
* converting palettes in `bokeh.palettes` to be python lists of integers

Other possibly useful feature:

* a python-side colormapper that can generate colors as numpy arrays of integers

With this, CDS columns of colors can be *much* more efficiently transferred.

**Comment by bryevdv** [comment]
tentatively moving this in to `2.0`. I'm not sure it will get done but if there is ever a time to consider it, this it it.
